training/src/main/scala/com/airbnb/aerosolve/training/pipeline/NDTreePipeline.scala

```diff
@@ -0,0 +1,194 @@
+import org.apache.spark.rdd.RDD
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Try
+
+object NDTreePipeline {
+  val log: Logger = LoggerFactory.getLogger("NDTreePipeline")
+
+  case class FeatureStats(count: Double, min: Double, max: Double)
+  /*
+    build NDTree from examples, each float/dense feature generates a NDTree
+    and save to FeatureMap
+    Sample config
+    make_feature_map {
+      input :  ${training_data}
+      output:  ${feature_map}
+      sample: 0.01
+      min_count: 200
+      max_tree_depth: 7  ( max nodes could be 2^(max_tree_depth-1) )
+      min_leaf_count: 200
+      // feature families in linear_feature should use linear
+      linear_feature: ["L", "T", "L_x_T"]
+    }
+    TODO add rankKey: LABEL (default to GenericPipeline.LABEL)
+  */
+  def buildFeatureMapRun(sc: SparkContext, config : Config) = {
+    val features = buildFeatures(sc, config)
+    for (((family, name), feature) <- features) {
+      // save to disk.
+      if (feature.isInstanceOf[FeatureStats]) {
+        val stats = feature.asInstanceOf[FeatureStats]
+        log.info(s"${family}, ${name}: ${stats.count} ${stats.min} ${stats.max}")
+      } else {
+        val model = feature.asInstanceOf[NDTreeModel]
+        val nodes = model.getNodes
+        log.info(s"${family}, ${name}: ${model.getDimension} ${nodes.length} ${nodes.mkString("\n")}")
+      }
+    }
+  }
+
+  def buildFeatures(sc: SparkContext, config : Config):
+      Array[((String, String), Any)] = {
+    val cfg = config.getConfig("make_feature_map")
+    val inputPattern: String = cfg.getString("input")
+    log.info(s"Training data: ${inputPattern}")
+    val sample: Double = cfg.getDouble("sample")
+    val minCount: Int = cfg.getInt("min_count")
+    // minMax.filter(x => !linearFeatureFamilies.contains(x._1._1))
+    val linearFeatureFamilies: java.util.List[String] = Try(config.getStringList("linear_feature"))
+      .getOrElse[java.util.List[String]](List.empty.asJava)
+
+    val input = GenericPipeline.getExamples(sc, inputPattern)
+    val options = NDTreeBuildOptions(
+      maxTreeDepth = cfg.getInt("max_tree_depth"),
+      minLeafCount = cfg.getInt("min_leaf_count"))
+
+    val result: Array[((String, String), Any)] = getFeatures(
+      sc, input, minCount, sample, linearFeatureFamilies, options)
+    result
+  }
+
+  def getFeatures(sc: SparkContext, input: RDD[Example], minCount: Int, sample: Double,
+                  linearFeatureFamilies: util.List[String],
+                  options: NDTreeBuildOptions): Array[((String, String), Any)] = {
+    val linearFeatureFamiliesBC = sc.broadcast(linearFeatureFamilies)
+    val optionsBC = sc.broadcast(options)
+    val tree: Array[((String, String), Any)] = input.sample(true, sample).mapPartitions(partition => {
```

Comment:
AFAIK, there is no union type in Scala out of the box. You can use `Either` type to make type more precise. Another idea is since feature family has low cardinality (you are collecting at the end) we could really aggregate/reduce with a Map (i.e. not partition by key at all).